File: Go/Minimum Depth of Binary Tree.go
// 4ms, 97.27%; 5.1MB, 100.00%
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func minDepth(root *TreeNode) int {
	if root == nil {
		return 0
	}
	cur, level := []*TreeNode{root}, 1
	for ; len(cur) > 0; level++ {
		tmp := make([]*TreeNode, 0, len(cur)<<1)
		for _, node := range cur {
			if node.Left == node.Right {
				return level
			}
			if node.Left != nil {
				tmp = append(tmp, node.Left)
			}
			if node.Right != nil {
				tmp = append(tmp, node.Right)
			}
		}
		cur = tmp
	}
	return -1 // unreachable
}

File: Java/Minimum Depth of Binary Tree.java
// 0ms, 100.00%; 39.6MB, 90.63%
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public static int minDepth(TreeNode root) {
        if (root == null) {
            return 0;
        }
        Deque<TreeNode> dq = new LinkedList<>();
        dq.offer(root);
        for (int i = 1;; ++i) {
            for (int j = dq.size(); j > 0; --j) {
                TreeNode node = dq.poll();
                if (node.left == node.right) {
                    return i;
                }
                if (node.left != null) {
                    dq.offer(node.left);
                }
                if (node.right != null) {
                    dq.offer(node.right);
                }
            }
        }
    }
}
